libkbfs/interfaces.go

```diff
@@ -1648,12 +1666,39 @@ type MutableBareRootMetadata interface {
 	// BareRootMetadata. This is necessary since newly-created
 	// BareRootMetadata objects don't have enough data to build a
 	// TlfHandle from until the first rekey.
-	FakeInitialRekey(h BareTlfHandle)
+	FakeInitialRekey(c Codec, h BareTlfHandle) (*TLFWriterKeyBundleV2, *TLFReaderKeyBundle, error)
 	// Update initializes the given freshly-created BareRootMetadata object with
 	// the given TlfID and BareTlfHandle. Note that if the given ID/handle are private,
 	// rekeying must be done separately.
 	// Update implements the BareRootMetadata interface for BareRootMetadataV2.
 	Update(tlf TlfID, h BareTlfHandle) error
 	// Returns the TLF key bundles for this metadata at the given key generation.
 	GetTLFKeyBundles(keyGen KeyGen) (*TLFWriterKeyBundle, *TLFReaderKeyBundle, error)
+	// GetUserDeviceKeyInfoMaps returns the given user device key info maps for the given
+	// key generation. Note "wkb" and "rkb" are expcted to be nil for pre-v3 metadata.
+	GetUserDeviceKeyInfoMaps(keyGen KeyGen, wkb *TLFWriterKeyBundleV2,
+		rkb *TLFReaderKeyBundle) (readers, writers UserDeviceKeyInfoMap, err error)
+	// fillInDevices ensures that every device for every writer and reader
+	// in the provided lists has complete TLF crypt key info, and uses the
+	// new ephemeral key pair to generate the info if it doesn't yet
+	// exist.
+	// Note "wkb" is expcted to be nil for post-v3 metadata and "wkb2" is expected to be
+	// nil for pre-v3 metadata.
+	fillInDevices(crypto Crypto,
+		wkb *TLFWriterKeyBundle, wkb2 *TLFWriterKeyBundleV2, rkb *TLFReaderKeyBundle,
```

Comment:
Having an `ExternalMetadata` interface would also avoid this awkwardness of having one parameter for one version and another for the other version.